File: backend/app/schemas/docdelta.py
"""Pydantic models for `POST /api/ai/docdelta`.

These models map 1:1 to `reference/doc_scheme.json` (the authoritative contract)
and are kept in sync with `frontend/src/types/docdelta.ts` manually
(packages/shared is a Phase 3 trigger — see T12).

Field names match the scheme byte-for-byte, including `new` (not a Python
reserved keyword — it is a built-in name but usable as an attribute).
`extra="forbid"` is applied so that unknown request fields are rejected at
the Pydantic layer, which is what the VALIDATION contract (422) prescribes.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class DocdeltaDocRef(BaseModel):
    """A single document reference. Used in `known_docs`, `new_doc`, `convert_doc`."""

    model_config = ConfigDict(extra="forbid")

    doc_id: str
    context: str


class DocdeltaRequest(BaseModel):
    """Request body for `POST /api/ai/docdelta`."""

    model_config = ConfigDict(extra="forbid")

    source_id: str
    instruction: str
    known_docs: list[list[DocdeltaDocRef]]
    new_doc: list[DocdeltaDocRef]
    convert_doc: list[DocdeltaDocRef] = Field(default_factory=list)


class DocdeltaConflict(BaseModel):
    """A single conflict entry in `output.conflict[]`."""

    model_config = ConfigDict(extra="forbid")

    doc_id: str
    known_text: str
    new_text: str
    reason: str
    severity: Literal["low", "medium", "high"]


class DocdeltaOutput(BaseModel):
    """The `output` object inside the response."""

    model_config = ConfigDict(extra="forbid")

    new: list[str]
    conflict: list[DocdeltaConflict]


class DocdeltaResponse(BaseModel):
    """200 OK response body for `POST /api/ai/docdelta`."""

    model_config = ConfigDict(extra="forbid")

    source_id: str
    output: DocdeltaOutput

File: frontend/src/types/docdelta.ts
/**
 * DocDelta API contract (frontend side).
 *
 * Authoritative source: `reference/doc_scheme.json`.
 * Shared with backend as a hand-maintained copy at `backend/src/types/docdelta.ts`
 * (no packages/shared until Phase 3 — see T12, _workspace/02_architect_decision.md §5).
 *
 * These types are a separate module from `src/types.ts` (the `Document` type).
 * `Document` is NOT modified — analysis results live only in component state,
 * never persisted into `Document` (hub T3: backward compatibility).
 */

/** A reference to a document passed to/through the docdelta endpoint. */
export interface DocdeltaDocRef {
  /** Maps from `Document.id`. */
  doc_id: string;
  /** Maps from `Document.content` (raw markdown, not HTML). */
  context: string;
}

/** Request body for `POST /api/ai/docdelta`. */
export interface DocdeltaRequest {
  /** Client-generated ad-hoc id per request. Format: `"S-" + crypto.randomUUID().slice(0, 8)`. */
  source_id: string;
  /** Fixed instruction string; server never overwrites. See `DOCDELTA_INSTRUCTION`. */
  instruction: string;
  /** 2D array of known-document groups. Phase 2 PoC always uses a single group: `[nonActiveDocs]`. */
  known_docs: DocdeltaDocRef[][];
  /** Exactly one element: the currently active document. */
  new_doc: DocdeltaDocRef[];
  /** Learning-schema compatibility field. PoC always sends `[]`; server tolerates it. */
  convert_doc: DocdeltaDocRef[];
}

/** Single conflict item from the response `output.conflict[]`. */
export interface DocdeltaConflict {
  /** Matches a `doc_id` from `known_docs` (any group). */
  doc_id: string;
  /** The conflicting passage as it exists in the known doc. */
  known_text: string;
  /** The conflicting passage as it appears in the new doc. */
  new_text: string;
  /** Human-readable rationale of the conflict. */
  reason: string;
  /** Normalized to one of three levels. */
  severity: "low" | "medium" | "high";
}

/** Response body for `POST /api/ai/docdelta` (200 OK, JSON mode). */
export interface DocdeltaResponse {
  /** Echoed from the request. */
  source_id: string;
  output: {
    /** New facts not present in any known doc. */
    new: string[];
    /** Conflicts between new_doc and known_docs. */
    conflict: DocdeltaConflict[];
  };
}

/** Error payload shape shared with the rest of the API (`docs/04_api.md` common format). */
export interface DocdeltaErrorResponse {
  error: {
    /** e.g. `"VALIDATION"`, `"AI_UPSTREAM"`, `"TIMEOUT"`. */
    code: string;
    message: string;
  };
}

/**
 * Fixed instruction string. Verbatim copy from `reference/doc_scheme.json`.
 * Must not be edited client-side in the PoC; user-editable instruction UI is a separate task.
 */
export const DOCDELTA_INSTRUCTION =
  "기존 시나리오/설정 문서들(known_docs)을 기준으로 업로드된 신규 시나리오 요약본(new_doc)에서 차이점을 분석하세요. new: 기존 설정에 없는 새로운 설정이나 시나리오 전개. conflict: 기존 설정이나 과거 시나리오와 논리적으로 충돌하는 내용. 단순한 표현 차이나 이미 존재하는 내용은 무시하고, 의미 있는 변화와 모순만 추출하세요.";
